sdk/eventhubs/azure_messaging_eventhubs/src/common/connection_manager.rs

```diff
@@ -0,0 +1,150 @@
+    url: Url,
+    application_id: Option<String>,
+    custom_endpoint: Option<Url>,
+    connections: Mutex<OnceLock<Arc<AmqpConnection>>>,
+    authorization_scopes: Mutex<HashMap<Url, AccessToken>>,
+}
+
+impl ConnectionManager {
+    pub fn new(url: Url, application_id: Option<String>, custom_endpoint: Option<Url>) -> Self {
+        Self {
+            url,
+            application_id,
+            custom_endpoint,
+            connections: Mutex::new(OnceLock::new()),
+            authorization_scopes: Mutex::new(HashMap::new()),
+        }
+    }
+
+    pub(crate) async fn ensure_connection(&self) -> Result<()> {
+        let connections = self.connections.lock().await;
+        if connections.get().is_some() {
+            return Ok(());
+        }
+
+        trace!("Creating connection for {}.", self.url);
+        let connection = Arc::new(AmqpConnection::new());
+        connection
+            .open(
+                self.application_id
+                    .clone()
+                    .unwrap_or(uuid::Uuid::new_v4().to_string()),
+                self.url.clone(),
+                Some(AmqpConnectionOptions {
+                    properties: Some(
+                        vec![
+                            ("user-agent", get_user_agent(&self.application_id)),
+                            ("version", get_package_version()),
+                            ("platform", get_platform_info()),
+                            ("product", get_package_name()),
+                        ]
+                        .into_iter()
+                        .map(|(k, v)| (AmqpSymbol::from(k), AmqpValue::from(v)))
+                        .collect(),
+                    ),
+                    custom_endpoint: self.custom_endpoint.clone(),
+                    ..Default::default()
+                }),
+            )
+            .await?;
+
+        trace!("Connection for {} created.", self.url);
+        connections
+            .set(connection)
+            .map_err(|_| EventHubsError::from(ErrorKind::UnableToAddConnection))?;
+        Ok(())
+    }
+
+    pub(crate) async fn get_connection(&self) -> Result<Arc<AmqpConnection>> {
+        let connections = self.connections.lock().await;
+
+        let connection = connections
+            .get()
+            .cloned()
+            .ok_or(EventHubsError::from(ErrorKind::MissingConnection))?;
+        Ok(connection)
+    }
+
+    pub(crate) async fn close_connection(&self) -> Result<()> {
+        let connections = self.connections.lock().await;
+        let connection = connections
+            .get()
+            .ok_or(EventHubsError::from(ErrorKind::MissingConnection))?;
+
+        connection.close().await?;
+        Ok(())
+    }
+
+    pub(crate) async fn authorize_path(
+        &self,
+        connection: &Arc<AmqpConnection>,
+        path: &Url,
+        credential: Arc<dyn azure_core::credentials::TokenCredential>,
+    ) -> Result<AccessToken> {
+        debug!("Authorizing path: {path}");
```

Comment:
I see you using `trace!()` in other places. Do you have, in your mind, a clear picture of `debug` vs `trace`? I know we haven't drafted guidelines yet (on my long todo list), but I'd suggest trying to at least be consistent in your lib. Not block, though.

Reply:
I use trace for the lowest level information - in other words, the things that *I* as the developer of the code care about. *debug* is used for things that I think that the *user* of the library would care about.

So when I'm sticking in "I need to make sure I understand what's going on here", I use `trace`, when I'm sticking in "I bet someone would like to know about this thing", I use `debug`.

But I'm not completely consistent, to be honest.